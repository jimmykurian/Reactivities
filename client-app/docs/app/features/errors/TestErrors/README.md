[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / app/features/errors/TestErrors

# app/features/errors/TestErrors

## Index

### Functions

- [default](functions/default.md)
