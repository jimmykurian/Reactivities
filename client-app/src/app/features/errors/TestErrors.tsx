/**
 * @file TestErrors.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the TestErrors component which is used to test error handling in the application.
 */

import { Button, Header, Segment } from 'semantic-ui-react';
import axios from 'axios';
import { useState } from 'react';
import ValidationError from './ValidationError';

/**
 * TestErrors component provides buttons to trigger various types of HTTP errors for testing purposes.
 * Each button triggers a specific HTTP request to test how the application handles different error responses.
 *
 * @component
 * @returns {JSX.Element} The rendered TestErrors component.
 */
export default function TestErrors() {
  const baseUrl = 'http://localhost:5000/api/';
  const [errors, setErrors] = useState(null);

  /**
   * Handles a 404 Not Found error by sending a GET request to a non-existent endpoint.
   */
  function handleNotFound() {
    axios
      .get(baseUrl + 'buggy/not-found')
      .catch((err) => console.log(err.response));
  }

  /**
   * Handles a 400 Bad Request error by sending a GET request that results in a client error.
   */
  function handleBadRequest() {
    axios
      .get(baseUrl + 'buggy/bad-request')
      .catch((err) => console.log(err.response));
  }

  /**
   * Handles a 500 Server Error by sending a GET request to trigger a server error.
   */
  function handleServerError() {
    axios
      .get(baseUrl + 'buggy/server-error')
      .catch((err) => console.log(err.response));
  }

  /**
   * Handles a 401 Unauthorized error by sending a GET request without proper authorization.
   */
  function handleUnauthorised() {
    axios
      .get(baseUrl + 'buggy/unauthorised')
      .catch((err) => console.log(err.response));
  }

  /**
   * Handles an error for a bad GUID format by sending a GET request with an invalid identifier.
   */
  function handleBadGuid() {
    axios
      .get(baseUrl + 'activities/notaguid')
      .catch((err) => console.log(err.response));
  }

  /**
   * Handles validation errors by sending a POST request with invalid data to trigger validation error responses.
   */
  function handleValidationError() {
    axios.post(baseUrl + 'activities', {}).catch((err) => setErrors(err));
  }

  return (
    <>
      <Header as="h1" content="Test Error component" />
      <Segment>
        <Button.Group widths="7">
          <Button onClick={handleNotFound} content="Not Found" basic primary />
          <Button
            onClick={handleBadRequest}
            content="Bad Request"
            basic
            primary
          />
          <Button
            onClick={handleValidationError}
            content="Validation Error"
            basic
            primary
          />
          <Button
            onClick={handleServerError}
            content="Server Error"
            basic
            primary
          />
          <Button
            onClick={handleUnauthorised}
            content="Unauthorised"
            basic
            primary
          />
          <Button onClick={handleBadGuid} content="Bad Guid" basic primary />
        </Button.Group>
      </Segment>
      {errors && <ValidationError errors={errors} />}
    </>
  );
}
