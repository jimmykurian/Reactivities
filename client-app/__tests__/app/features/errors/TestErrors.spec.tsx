import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import axios from 'axios';
import TestErrors from '../../../../src/app/features/errors/TestErrors';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('TestErrors Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('handles Not Found error correctly', async () => {
    // Arrange
    mockedAxios.get.mockRejectedValueOnce({
      response: { status: 404, data: 'Not Found' },
    });
    render(<TestErrors />);

    // Act
    fireEvent.click(screen.getByText('Not Found'));

    // Assert
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'http://localhost:5000/api/buggy/not-found',
      ),
    );
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  it('handles Bad Request error correctly', async () => {
    // Arrange
    mockedAxios.get.mockRejectedValueOnce({
      response: { status: 400, data: 'Bad Request' },
    });
    render(<TestErrors />);

    // Act
    fireEvent.click(screen.getByText('Bad Request'));

    // Assert
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'http://localhost:5000/api/buggy/bad-request',
      ),
    );
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  it('handles Server Error correctly', async () => {
    // Arrange
    mockedAxios.get.mockRejectedValueOnce({
      response: { status: 500, data: 'Server Error' },
    });
    render(<TestErrors />);

    // Act
    fireEvent.click(screen.getByText('Server Error'));

    // Assert
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'http://localhost:5000/api/buggy/server-error',
      ),
    );
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  it('handles Unauthorised error correctly', async () => {
    // Arrange
    mockedAxios.get.mockRejectedValueOnce({
      response: { status: 401, data: 'Unauthorised' },
    });
    render(<TestErrors />);

    // Act
    fireEvent.click(screen.getByText('Unauthorised'));

    // Assert
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'http://localhost:5000/api/buggy/unauthorised',
      ),
    );
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  it('handles Bad Guid error correctly', async () => {
    // Arrange
    mockedAxios.get.mockRejectedValueOnce({
      response: { status: 400, data: 'Bad Guid' },
    });
    render(<TestErrors />);

    // Act
    fireEvent.click(screen.getByText('Bad Guid'));

    // Assert
    await waitFor(() =>
      expect(mockedAxios.get).toHaveBeenCalledWith(
        'http://localhost:5000/api/activities/notaguid',
      ),
    );
    expect(mockedAxios.get).toHaveBeenCalledTimes(1);
  });

  it('renders all buttons', () => {
    // Arrange & Act
    const { container } = render(<TestErrors />);

    // Assert
    expect(screen.getByText('Not Found')).toBeInTheDocument();
    expect(screen.getByText('Bad Request')).toBeInTheDocument();
    expect(screen.getByText('Validation Error')).toBeInTheDocument();
    expect(screen.getByText('Server Error')).toBeInTheDocument();
    expect(screen.getByText('Unauthorised')).toBeInTheDocument();
    expect(screen.getByText('Bad Guid')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
