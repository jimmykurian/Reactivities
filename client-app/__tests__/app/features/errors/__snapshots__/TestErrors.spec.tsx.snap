// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestErrors Component renders all buttons 1`] = `
<div>
  <h1
    class="ui header"
  >
    Test Error component
  </h1>
  <div
    class="ui segment"
  >
    <div
      class="ui seven buttons"
    >
      <button
        class="ui basic primary button"
      >
        Not Found
      </button>
      <button
        class="ui basic primary button"
      >
        Bad Request
      </button>
      <button
        class="ui basic primary button"
      >
        Validation Error
      </button>
      <button
        class="ui basic primary button"
      >
        Server Error
      </button>
      <button
        class="ui basic primary button"
      >
        Unauthorised
      </button>
      <button
        class="ui basic primary button"
      >
        Bad Guid
      </button>
    </div>
  </div>
</div>
`;
